RPC: RPCConnections can now handle requests in both directions

omerfirmak opened the pull request:
No concurrent request in the same direction.

omerfirmak committed:
RPC: RPCConnections can now handle requests in both directions

No concurrent request in the same direction.

omerfirmak committed:
RPC: Move RPCConnection to module scope

We need to be able to create them without being bound
to a RPCClient.

omerfirmak committed:
RPC: make RPCConnection disposable

omerfirmak commented:
@Geod24 , I can't push to our `vibe-core` fork.

Geod24 commented:
@omerfirmak : Fixed

Geod24 commented on source/agora/network/RPC.d:
```diff
@@ -298,6 +289,12 @@ private class RPCConnection
     alias conn this;
+
+    ///
+    this (TCPConnection conn) @safe
```
Attributes. Especially `nothrow` because the compiler generates much more efficient code for `nothrow` ctors.

Geod24 commented on source/agora/network/Manager.d:
```diff
@@ -1101,7 +1105,7 @@ public class NetworkManager
     ***************************************************************************/
 
-    protected API getClient (Address address, Duration timeout)
+    protected agora.api.Validator.API getClient (Address address, Duration timeout)
```
Why does this return a `Validator` API ? What happens for full nodes ?

omerfirmak replied:
`API` was always `Validator.API` , we just have to specify it explicitly now that I have also imported `FullNode.API`.

> What happens for full nodes ?

Nothing, we just don't call `Validator` only endpoints on them.

Geod24 reviewed (commented):
I see the main problem is that it creates a dependency between the `NetworkManager` and the calling node. Will take a deeper look tomorrow after the release.

omerfirmak commented:
> I see the main problem is that it creates a dependency between the NetworkManager and the calling node. Will take a deeper look tomorrow after the release.

Yeah, I don't like that either. I'll update it if I find a cleaner solution.

omerfirmak committed:
Manager/RPC: Allow new RPC connections to be merged to existing ones

When a remote initiates a new RPC connection, we will try to discover them
as an Agora node. In the case where the remote is actually an Agora node, NetworkManager
will try to merge that connection into an existing RPC connection. This way incoming connections
can be used for outgoing requests as well.

omerfirmak commented:
Updated with discovery thru incoming RPC conns. Integration is happy.

Geod24 commented:
> RPC: Remove RPC client lookup table

Why ?

Geod24 commented:
Please upstream the vibe-core fixes.

omerfirmak commented:
> 
> 
> > RPC: Remove RPC client lookup table
> 
> Why ?

Dropped it.

> Please upstream the vibe-core fixes.

Will do

Geod24 commented on source/agora/network/Manager.d:
```diff
@@ -129,6 +133,19 @@ public class NetworkManager
             this.address = address;
+            this.api = null;
+            this.onHandshakeComplete = onHandshakeComplete;
+            this.onFailedRequest = onFailedRequest;
```
```suggestion
            this(null, address, onHandshakeComplete, onFailedRequest);
```

Geod24 commented on source/agora/network/Manager.d:
```diff
@@ -129,6 +133,19 @@ public class NetworkManager
+        public this (agora.api.Validator.API api,
+            void delegate (scope ref NodeConnInfo node) onHandshakeComplete,
+            bool delegate (in Address address) onFailedRequest,
+            Address address = Address.init)
```
Do we need a default value ?
If not, keep the parameters in the same order (e.g. `address` after `api`).

Geod24 reviewed (approved):
One last thing

Geod24 commented:
Rebased it and addressed my comments.

omerfirmak commented:
So this is back

```
[source/agora/test/Base.d:1339] Assertion thrown during test: Trying to access node at address 'Validator-1.boa1xzval2a3cdxv28n6slr62wlczslk3juvk7cu05qt3z55ty2rlfqfc6egsh2.localrest' from 'Validator-1.boa1xzval2a3cdxv28n6slr62wlczslk3juvk7cu05qt3z55ty2rlfqfc6egsh2.localrest' without first creating it
source/agora/test/Base.d:1339 agora.test.Base.TestAPI agora.test.Base.TestNetworkManager.getClient(immutable(char)[], core.time.Duration) [0x103c58d9f]
source/agora/network/Manager.d:186 void agora.network.Manager.NetworkManager.ConnectionTask.connect_canthrow() [0x103d97471]
source/agora/network/Manager.d:177 nothrow void agora.network.Manager.NetworkManager.ConnectionTask.connect() [0x103d9735b]
core/thread/context.d:46 void core.thread.context.Callable.opCall() [0x104a2bb9c]
core/thread/fiber.d:1141 void core.thread.fiber.Fiber.run() [0x104a2bd28]
core/thread/fiber.d:198 fiber_entryPoint [0x104a2bc65]
#### FATAL ERROR: Trying to access node at address 'Validator-1.boa1xzval2a3cdxv28n6slr62wlczslk3juvk7cu05qt3z55ty2rlfqfc6egsh2.localrest' from 'Validator-1.boa1xzval2a3cdxv28n6slr62wlczslk3juvk7cu05qt3z55ty2rlfqfc6egsh2.localrest' without first creating it
This node was started at source/agora/test/BlockRewards.d:45
This most likely means that the node crashed due to an uncaught exception
If not, please file a bug at https://github.com/Geod24/localrest/
Full error: core.exception.AssertError@source/agora/test/Base.d(1339): Trying to access node at address 'Validator-1.boa1xzval2a3cdxv28n6slr62wlczslk3juvk7cu05qt3z55ty2rlfqfc6egsh2.localrest' from 'Validator-1.boa1xzval2a3cdxv28n6slr62wlczslk3juvk7cu05qt3z55ty2rlfqfc6egsh2.localrest' without first creating it
```

Geod24 commented:
Yeah, but the worrying part is integration: Maybe my rebase was incorrect, as it seems to always re-establish connections.

Geod24 commented:
Also, it's a node trying to connect to itself. They should ban themselves. I'll take a look.

omerfirmak commented:
> Yeah, but the worrying part is integration: Maybe my rebase was incorrect, as it seems to always re-establish connections.

That wasnt the case, I will take a look.

omerfirmak commented:
Ok, I see what is happening. Incoming connections from FullNodes cannot be merged as they can not be identified by a key. So every time a FullNode initiates a connections, that is registered as a different peer. Which easily maxes the listener limit of a node.

omerfirmak commented:
Also, now that we have both TCP and HTTP endpoints, they both count as a separate peer and take up a peer slot.

Geod24 commented:
Should / can we prioritize the tcp connection in that situation ?

Geod24 commented:
> Ok, I see what is happening. Incoming connections from FullNodes cannot be merged as they can not be identified by a key. So every time a FullNode initiates a connections, that is registered as a different peer. Which easily maxes the listener limit of a node.

Why does it keeps on initiating connections tho ?
